docs/guide/ethereum-provider.md

```diff
@@ -112,83 +173,359 @@ params: [
+The kind of message is identified by the `type` string.
+
+One prominent example of such messages include updates from RPC subscriptions, such as `eth_subscribe`. For such messages, the message `type` will be `eth_subscription`.
+
+## Errors
+
+All errors thrown or returned by the MetaMask provider follow this interface:
+
+```typescript
+interface ProviderRpcError extends Error {
+  message: string;
+  code: number;
+  data?: unknown;
+}
+```
+
+The [`ethereum.request(args)` method](#ethereum-request-args) throws errors eagerly.
+You can often use the error `code` property to determine why the request failed.
+Common codes and their meaning include:
+
+- `4001`
+  - The request was rejected by the user
+- `-32602`
+  - The parameters were invalid
+- `-32603`
+  - Internal error
+
+For the complete list of errors, please see [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193#provider-errors) and [EIP-1474](https://eips.ethereum.org/EIPS/eip-1474#error-codes).
+
+::: tip
+The [`eth-rpc-errors`](https://npmjs.com/package/eth-rpc-errors) package implements all RPC errors thrown by the MetaMask provider, and can help you identify their meaning.
+:::
+
+## Using the Provider
+
+This snippet explains how to accomplish the three most common requirements for web3 sites:
+
+- Detect the Ethereum provider (`window.ethereum`)
+- Detect which Ethereum network the user is connected to
+- Get the user's Ethereum account(s)
+
+<<< @/docs/snippets/handleProvider.js
+
+## Experimental API
 
 ::: warning
-Use `ethereum.sendAsync({ method: 'eth_requestAccounts' }, callback)` instead.
+There is no guarantee that the methods and properties defined in this section will remain stable.
+Use it at your own risk.
 :::
 
-Requests to view the user's Ethereum address.
-Returns a promise of an array of hex-prefixed ethereum address strings.
+We expose some experimental, MetaMask-specific methods under the `ethereum._metamask` property.
+
+## Experimental Methods
+
+### ethereum.\_metamask.isApproved() (TO BE REMOVED)
+
+::: danger DANGER
+This will be removed in **Q3 2020**.
+:::
+
+```typescript
+ethereum._metamask.isApproved(): Promise<boolean>;
+```
+
+This method returns a `Promise` that resolves to a `boolean` indicating if the caller has access to user accounts.
 
-### ethereum.send(payload, callback) (DEPRECATED)
+### ethereum.\_metamask.isEnabled() (TO BE REMOVED)
+
+::: danger DANGER
+This will be removed in **Q3 2020**.
+:::
+
+```typescript
+ethereum._metamask.isEnabled(): boolean;
+```
+
+This method returns a `boolean` indicating if the caller has access to user accounts.
+
+### ethereum.\_metamask.isUnlocked()
+
+```typescript
+ethereum._metamask.isUnlocked(): Promise<boolean>;
+```
+
+This method returns a `Promise` that resolves to a `boolean` indicating if MetaMask is unlocked by the user.
+This is useful for knowing if MetaMask is unlocked in order to provide meaningful instructions to the user during onboarding.
```

Comment:
During onboarding specifically? This seems useful generally all of the time, not just during onboarding. The dapp can't do much if the extension is locked.

Reply:
3d130b6